rclrs/src/context.rs

```diff
@@ -142,6 +150,24 @@ impl Context {
             rcl_init_options_fini(&mut rcl_init_options).ok()?;
             // Move the check after the last fini()
             ret?;
+
+            // TODO: "Auto set-up logging" is forced but should be configurable as per rclcpp and rclpy
+            // SAFETY:
+            // * Lock the mutex as we cannot guarantee that rcl_* functions are protecting their global variables
+            // * Context is validated before we reach this point
+            // * No other preconditions for calling this function
+            let ret = {
+                let _lifecycle_lock = ENTITY_LIFECYCLE_MUTEX.lock().unwrap();
+                // Note: shadowing the allocator above.  The rcutils_get_default_allocator provides us with mechanisms for allocating and freeing
+                // memory, e.g. calloc/free. As these are function pointers and will be the same each time we call the method, it is safe to
+                // perform this shadowing
+                // Alternate is to call rcl_init_options_get_allocator however, we've freed the allocator above and restructuring the call
+                // sequence is unnecessarily complex
+                let allocator = rcutils_get_default_allocator();
+
+                rcl_logging_configure(&rcl_context.global_arguments, &allocator).ok()
```

Comment:
This will get called every time a Context is created, but it's meant to be called only once. It's possible for users to create multiple Contexts in one application (we do this a lot in tests), so we'd be making redundant calls.

Similar to https://github.com/ros2/rclcpp/pull/998 we should guard this from being called multiple times. We also need to make sure that `rcl_logging_fini` is only called after all Contexts are destructed.

This is a tricky problem to resolve in a sound way. I might suggest something like this:
```
struct LoggingConfiguration {
    logging: Mutex<Weak<LoggingLifecycle>>
}

impl LoggingConfiguration {
    fn configure(args: &rcl_arguments_t) -> Result<Arc<LoggingLifecycle>, RclError> {
        static CONFIGURATION: LazyLock<Self> = LazyLock::new(|| Self { logging: Mutex::new(Weak::new()) });
        let mut logging = CONFIGURATION.logging.lock().unwrap();
        if let Some(arc_logging) = logging.upgrade() {
            return Ok(arc_logging);
        }
        let arc_logging = Arc::new(LoggingLifecycle::new(args)?);
        *logging = arc_logging.downgrade();
        Ok(arc_logging);
    }
}

struct LoggingLifecycle;
impl LoggingLifecycle {
    fn new(args: &rcl_arguments_t) -> Result<Self, RclError> {
        let _lifecycle_lock = ENTITY_LIFECYCLE_MUTEX.lock().unwrap();
        let allocator = rcutils_get_default_allocator();
        rcl_logging_configure(args, &allocator).ok()?;
        Self
    }
}

impl Drop for LoggingLifecycle {
    fn drop(&mut self) {
        let _lifecycle_lock = ENTITY_LIFECYCLE_MUTEX.lock().unwrap();
        rcl_logging_fini();
    }
}
```
Then we would add an `Arc<LoggingLifecycle>` to the `ContextHandle`. Once all `ContextHandle`s are gone, the `LoggingLifecycle` will drop and call `rcl_logging_fini`. If a new context is made after that, `rcl_logging_configure` will be called again.

Reply:
I took the liberty of opening a PR for this: https://github.com/geoff-imdex/ros2_rust/pull/6

It's _almost_ exactly what I recommended above.